src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/NamedPipeBasedRazorProjectInfoDriver.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.Threading;
+
+#if !NET 
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+#endif
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+internal sealed class NamedPipeBasedRazorProjectInfoDriver : AbstractRazorProjectInfoDriver
+{
+    private NamedPipeClientStream? _namedPipe;
+
+    public NamedPipeBasedRazorProjectInfoDriver(ILoggerFactory loggerFactory) : base(loggerFactory)
+    {
+        StartInitialization();
+    }
+
+    public async Task ConnectAsync(string pipeName, CancellationToken cancellationToken)
+    {
+        Logger.LogTrace($"Connecting to named pipe {pipeName}");
+        Assumed.True(_namedPipe is null);
+
+#if NET
+        _namedPipe = new NamedPipeClientStream(".", pipeName, PipeDirection.In, PipeOptions.CurrentUserOnly | PipeOptions.Asynchronous);
+#else
+        _namedPipe = new NamedPipeClientStream(".", pipeName, PipeDirection.In, PipeOptions.Asynchronous);
+        CheckPipeConnectionOwnership(_namedPipe);
+#endif
+
+        await _namedPipe.ConnectAsync(cancellationToken).ConfigureAwait(false);
+        ReadFromStreamAsync(CancellationToken.None).Forget();
+    }
+
+    protected override Task InitializeAsync(CancellationToken cancellationToken) => Task.CompletedTask;
+
+    protected override void OnDispose()
+    {
+        _namedPipe?.Dispose();
+        _namedPipe = null;
+    }
+
+    private async Task ReadFromStreamAsync(CancellationToken cancellationToken = default)
+    {
+        _namedPipe.AssumeNotNull();
```

Comment:
Why do  we use this pattern vs. `Debug.Assert(_namedPipe is not null)`?

Reply:
Preference on exception. Something is really wrong if the method is started but we do not have a named pipe.